File: Data Structure & Algorithm/Models/Binary Search/Find Upper bound/Binary Search.ts
// Find Upper bound
export function binarySearch(nums: number[], target: number): number {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    const mid = left + Math.floor((right - left) / 2);
    if (nums[mid] <= target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  if (left > 0 && nums[left - 1] === target) {
    return left - 1;
  } else {
    return -1;
  }
}

// test
const res = binarySearch([-1, 0, 3, 5, 9, 12], 9);
console.log(res);

File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Template/Binary Search/Find Upper bound/Binary Search.go
package main

import "fmt"

// Find Upper bound
func binarySearch(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] >= target {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}

	if right+1 < len(nums) && nums[right+1] == target {
		return right + 1
	}

	return -1
}

func main() {
	nums := []int{-1, 0, 3, 5, 9, 12}
	target := 9
	res := binarySearch(nums, target)
	fmt.Println(res)
}
